sql/src/main/java/io/druid/sql/calcite/rel/QueryMaker.java

```diff
@@ -95,64 +95,88 @@ public Void apply(final Row row)
     };
   }
 
-  public static void executeSelect(
-      final DruidTable druidTable,
+  public void accumulate(
+      final DataSource dataSource,
+      final RowSignature sourceRowSignature,
+      final DruidQueryBuilder queryBuilder,
+      final Function<Row, Void> sink
+  )
+  {
+    if (dataSource instanceof QueryDataSource) {
+      final GroupByQuery outerQuery = queryBuilder.toGroupByQuery(dataSource, sourceRowSignature);
+      if (outerQuery == null) {
+        // Bug in the planner rules. They shouldn't allow this to happen.
+        throw new IllegalStateException("Can't use QueryDataSource without an outer groupBy query!");
+      }
+
+      executeGroupBy(queryBuilder, outerQuery, sink);
+      return;
+    }
+
+    final TimeseriesQuery timeseriesQuery = queryBuilder.toTimeseriesQuery(dataSource, sourceRowSignature);
+    if (timeseriesQuery != null) {
+      executeTimeseries(queryBuilder, timeseriesQuery, sink);
```

Comment:
I think it would be better if we are able to know the query type ahead. But, I know this ```accumulate()``` method is just moved from ```DruidQueryBuilder``` with little changes, and adding query types will involve a lot of additional changes. Do you have any plan for this?